src/com/google/javascript/jscomp/ConvertChunksToESModules.java

```diff
@@ -0,0 +1,192 @@
+        Node exportSpec = new Node(Token.EXPORT_SPEC);
+        exportSpec.addChildToFront(IR.name(name));
+        exportSpec.addChildToFront(IR.name(name));
+        exportSpec.putIntProp(Node.IS_SHORTHAND_PROPERTY, 1);
+        exportSpecs.addChildToBack(exportSpec);
+      }
+      Node export = IR.export(exportSpecs).useSourceInfoFromForTree(astRoot);
+      moduleBody.addChildToBack(export);
+      compiler.reportChangeToChangeScope(astRoot);
+    }
+
+    // Add import statements to chunks
+    for (Map.Entry<JSModule, Map<JSModule, Set<String>>> jsModuleImports :
+        crossChunkImports.entrySet()) {
+      ArrayList<Node> importStatements = new ArrayList<>();
+      CompilerInput firstInput = jsModuleImports.getKey().getInput(0);
+      Node astRoot = firstInput.getAstRoot(compiler);
+      Node moduleBody =
+          astRoot.getFirstChild() != null && astRoot.getFirstChild().isModuleBody()
+              ? astRoot.getFirstChild()
+              : null;
+      if (moduleBody == null) {
+        FeatureSet scriptFeatures = NodeUtil.getFeatureSetOfScript(astRoot);
+        scriptFeatures = scriptFeatures.union(FeatureSet.ES6_MODULES);
+        astRoot.putProp(Node.FEATURE_SET, scriptFeatures);
+        moduleBody = new Node(Token.MODULE_BODY);
+        moduleBody.useSourceInfoFrom(astRoot);
+        moduleBody.addChildrenToFront(astRoot.removeChildren());
+        astRoot.addChildToFront(moduleBody);
+      }
+      for (Map.Entry<JSModule, Set<String>> jsModuleImportsFromChunk :
+          jsModuleImports.getValue().entrySet()) {
+        Node importSpecs = new Node(Token.IMPORT_SPECS);
+        for (String name : jsModuleImportsFromChunk.getValue()) {
+          Node importSpec = new Node(Token.IMPORT_SPEC);
+          importSpec.addChildToFront(IR.name(name));
+          importSpec.addChildToFront(IR.name(name));
+          importSpec.putIntProp(Node.IS_SHORTHAND_PROPERTY, 1);
+          importSpecs.addChildToBack(importSpec);
+        }
+        Node importStatement = new Node(Token.IMPORT);
+        String importPath =
+            Compiler.resolveSibling(
+                jsModuleImports.getKey().getName(),
+                jsModuleImportsFromChunk.getKey().getName() + ".js");
+        if (!importPath.startsWith(".") && !importPath.startsWith("/")) {
+          importPath = "./" + importPath;
+        }
+        importStatement.addChildToFront(IR.string(importPath));
+        importStatement.addChildToFront(importSpecs);
+        importStatement.addChildToFront(IR.empty());
+        importStatement.useSourceInfoFromForTree(astRoot);
+        importStatements.add(0, importStatement);
+      }
+      for (Node importStatement : importStatements) {
+        moduleBody.addChildToFront(importStatement);
+      }
+      compiler.reportChangeToChangeScope(astRoot);
+    }
+  }
+
+  /** Find names in a module that are defined in a different module. */
+  private class FindCrossChunkReferences extends AbstractPostOrderCallback {
```

Comment:
We'd never encourage people to write code this way, but perhaps the simplest thing we could do to break our big output script into ES-module-style chunks would be to do everything as side effects rather than trying to track all the variables shared across chunks.

If we required `--isolation_mode=IIFE` and `--rename_prefix_namespace=_` or similar, then we could just not bother generating any exports and use the JSModuleGraph to generate the appropriate imports for each chunk. It's not *quite* that simple, but it might still be better than generating potentially enormous import statements and export statements.

 WDYT?

Reply:
If you chain RescopeGlobalSymbols and this pass, that's almost what you get.

However that's definitely not my goal. I'm trying to get proper ES modules as output and the single namespace export that you pass around using imports / exports violates the spirit of that. I want to run this on some real world projects, but I believe the potentially enormous import/export statements are going to be the minority of projects. I don't believe that will be a problem for the majority of the open source world.

I would also expect a pretty strong negative developer reaction to the namespace idea. I'm wanting to fully embrace ES module output here.

Reply:
OK, I can see that.

So, do you have a real-world project handy you can try it on?

Reply:
Sure do!